Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,447 @@
+  where
+    proof := by
+      simp only [size_swap]
+      exact v.size_eq
+      done
+
+/--
+Swaps two entries in an vector, or panics if either index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def swap! (v : Vector α n) (i j : Nat) : Vector α n :=
+  ⟨Array.swap! v.toArray i j, by {simp only [size_swap!]; exact v.size_eq}⟩
+
+/--
+Swaps the entry with index `i : Fin n` in the vector for a new entry.
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def swapAt (v : Vector α n) (i : Fin n) (x : α) : α × Vector α n:=
+  let e := Vector.get v i
+  let v := Vector.set v i x
+  (e, v)
+
+
+instance [Inhabited α] : Inhabited (Vector α n) where
+  default := mkVector n default
+
+/--
+Swaps the entry at index `i : Nat`in the vector for a new entry, if the index is valid.
+Otherwise it panics
+The old entry is returned with the modified vector.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def swapAt! (v : Vector α n) (i : Nat) (x : α) : α × Vector α n :=
+  if h : i < n then
+    swapAt v ⟨i, h⟩ x
+  else
+    have : Inhabited α := ⟨x⟩
+    panic! ("index " ++ toString i ++ " out of bounds")
+
+/--Checks whether the vector is empty, that is, its size is `0`-/
+def isEmpty (_ : Vector α n) : Bool :=
+  n = 0
+
+/--Returns a vector `#v[1,2,3,...,n-1]`-/
+def range (n : Nat) : Vector Nat n :=
+  ⟨Array.range n, by {simp}⟩
+
+/--
+Returns a vector of size `1` with a single element `v`
+-/
+def singleton (v : α) : Vector α 1 :=
+  mkVector 1 v
+
+/--
+Vector lookup function that takes an index of type `USize`
+-/
+def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α := Vector.get v ⟨i.toNat, h⟩
+
+/--
+borrowed from array API:
+auxiliary declaration used in the equation compiler when pattern matching array literals.
+-/
+abbrev getLit {α : Type u} {n : Nat} (v : Vector α n) (i : Nat)
+  (h₁ : Vector.size v = n) (h₂ : i < n) : α :=
+  have := h₁.symm ▸ h₂
+  Vector.get v ⟨i, h₂⟩
+
+instance : GetElem (Vector α n) Nat α fun _ i => i < n where
+  getElem := fun x i h => get x ⟨i, h⟩
+
+/--
+`pop v` returns the vector with the last element removed
+-/
+def pop (v : Vector α n) : Vector α (n-1) :=
+  ⟨Array.pop v.toArray, by {simp only [size_pop, v.size_eq]}⟩
+
+/--
+`shrink v m` removess the last `m` elements of a vector.
+-/
+def shrink (v : Vector α n) (m : Nat) : Vector α (n - m) :=
+  match m with
+  | 0 => v
+  | (x + 1) =>
+      have h : n - 1 - x = n - (x + 1) := by
+        omega
+      h ▸ (shrink (Vector.pop v) x)
+
+/--
+`isEqvAux a b hsz p i` is a helper function for `isEqv`. It performs
+an index by index check on whether the corresponding elements of a vector
+satisfy the boolean property `p` starting at a given index `i`
+-/
+def isEqvAux (a b : Vector α n)
+  (hsz : size a = size b)
```

Comment:
That hypothesis is just `n = n`. I don't think this def is needed for vectors.

Reply:
This is an aux theorem because I had issues with `apply`ing `Array.ext`

Reply:
Oh shoot I see what you mean

Reply:
Done